src/cocotb/__init__.py

```diff
@@ -241,22 +238,40 @@ def create_task(
         )
 
 
-def _initialise_testbench(argv_):  # pragma: no cover
-    try:
-        _start_library_coverage()
-        _initialise_testbench_(argv_)
-    except BaseException:
-        log.exception("cocotb testbench initialization failed. Exiting.")
-        from cocotb import simulator
+_shutdown_callbacks: List[Callable[[], None]] = []
+"""List of callbacks to be called when cocotb shuts down."""
+
+
+def _register_shutdown_callback(cb: Callable[[], None]) -> None:
+    """Register a callback to be called when cocotb shuts down."""
+    _shutdown_callbacks.append(cb)
+
 
-        simulator.stop_simulator()
-        _stop_library_coverage()
+def _shutdown_testbench() -> None:
+    """Call all registered shutdown callbacks."""
+    for cb in _shutdown_callbacks:
+        cb()
 
 
-def _initialise_testbench_(argv_):
+def _initialise_testbench(argv_: List[str]) -> None:
     # The body of this function is split in two because no coverage is collected on
     # the function that starts the coverage. By splitting it in two we get coverage
     # on most of the function.
+    try:
+        _initialise_testbench_(argv_)
+    except BaseException:
+        print("cocotb testbench initialization failed. Exiting.", file=sys.stderr)
+        _shutdown_testbench()
+
+    # start Regression Manager
+    regression_manager.start_regression()
```

Comment:
This is outside the above `try` block, so when the `KeyboardInterrupt` is raised in the test, it ends up calling the registered sim event callback, which fails the test and the next one in the RegressionManager.

Reply:
Yeah I forgot the results.xml is still written for these tests and will fail without setting expect_sim_failure.